### DISCUSSION TITLE
Wix 4 conversion: RegistryValue under Component

### DISCUSSION THREAD

**Discussion by wmanning** [wixtoolset/issues#7336] [Questions]
I have a v3.11 wxs file containing:

```xml
<?xml version="1.0" encoding="UTF-8"?>
 <Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
     	<Fragment>
           <ComponentGroup Id="CMG_SomeComponentGroup">
               <ComponentRef Id="CMP_SomeComponent" />
          </ComponentGroup>

           <DirectoryRef Id="TARGETDIR">
              <Component Id="CMP_SomeComponent" Guid="{some-guid}">
                  <RegistryValue Root="HKLM"
                         Key="Software\some-company\some-product\Addons"
                         Name="some-name"
                         Type="string"
                         Value=""
                         KeyPath="yes" />
              </Component>
         </DirectoryRef>
 </Wix>
```

After converting to Wix 4.0 it became:

```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
   <Fragment>
     <ComponentGroup Id="CMG_SomeComponentGroup">
          <ComponentRef Id="CMP_SomeComponent" />
     </ComponentGroup>

      <Component Id="CMP_SomeComponent" Guid="{some-guid}">
          <RegistryValue Root="HKLM" Key="Software\some-company\some-product\Addons" Name="some-name" Type="string" Value="" KeyPath="yes" />
     </Component>
   </Fragment>
</Wix>
```

I get following error in the log (which is understandable):
>  *** ERROR ***: someWixFile.wxs : error WIX0070: A reference to the TARGETDIR Directory was removed. This may cause the Fragment that defined TARGETDIR to not be included in the final output. If this happens, reference a different element in the Fragment to replace the old reference to TARGEDIR. (TargetDirRefRemoved)

When I build I get the following build error:

> error WIX0010: The Component/@Directory attribute was not found; it is required.

Which makes sense as the Component no longer is wrapped in the ```<DirectoryRef Id="TARGETDIR">``` element.  

I wrap the component in a directoryref pointed to my install folder ("some-drive/programsFolder/some-manufacturer/some-product") and problem solved.

Should I create an issue to request that the conversion log mention specifically that it removed the directoryRef and the user should add one back in or maybe that it shouldn't delete the directoryref and just delete the TARGETDIR text within the Id attribute and ask the user to refer to another directory ID?

The above component contains a registeryValue and not a file so would it be a good feature request for WixToolset to automatically add a directoryref to that component during build?  The same way (I think) that the StandardDirectory element is creating a ``` <Directory Id="TARGETDIR" Name="">``` in the background.

Or is there a reason a developer has to specify a directory for a component only containing a registry value?

**Comment by barnson** [comment]
All components have a directory, regardless of what's in them. TARGETDIR can change between install and repair/uninstall/patch/upgrade, so it's not a good directory to use. Typically, `<DirectoryRef Id="TARGETDIR">` has other `Directory` children, which would still work.

A better message would be great.

**Reply by wmanning** [reply]
> TARGETDIR can change between install and repair/uninstall/patch/upgrade

Good to know, thanks. I will change this. In my case I do have other directories under ```<DirectoryRef Id="TARGETDIR">``` in my main wxs.

I will create an issue about the message.

**Comment by robmen** [comment]
Also, you can simplify your code both in WiX v3 and v4 by doing the following:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
     	<Fragment>
           <ComponentGroup Id="CMG_SomeComponentGroup" Directory="PICK_SOMETHING_BETTER_THAN_TARGETDIR">
              <Component Id="CMP_SomeComponent" Guid="{some-guid}">
                  <RegistryValue Root="HKLM"
                         Key="Software\some-company\some-product\Addons"
                         Name="some-name"
                         Type="string"
                         Value=""
                         KeyPath="yes" />
              </Component>
          </ComponentGroup>
</Wix>
```

And when you can change the GUID you can simplify it to:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
     	<Fragment>
           <ComponentGroup Id="CMG_SomeComponentGroup" Directory="PICK_SOMETHING_BETTER_THAN_TARGETDIR">
              <Component>
                  <RegistryValue Root="HKLM"
                         Key="Software\some-company\some-product\Addons"
                         Name="some-name"
                         Type="string"
                         Value="" />
              </Component>
          </ComponentGroup>
</Wix>
```

Cleaner code will convert better.

**Reply by wmanning** [reply]
Thanks. I like keeping the component group and component separate  to have a quick summary of what components are under what componentRef.  Doesn't apply to this one but like to do that in general.

**Reply by robmen** [reply]
Interesting. Lots more typing, IMHO, but there is no accounting for taste. 😜

FYI, the overall goal is to make `TARGETDIR` disappear. It generally should not be used because it has side-effects by changing to drive with the most free space. So, there is a little pain while we try to find ways through this transition.
